File: Avail/src/com/avail/server/io/AvailServerChannel.java
/**
 * AvailServerChannel.java
 * Copyright © 1993-2015, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.server.io;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicLong;
import com.avail.annotations.Nullable;
import com.avail.io.TextInterface;
import com.avail.server.AvailServer;
import com.avail.server.messages.CommandMessage;
import com.avail.server.messages.Message;
import com.avail.utility.evaluation.Continuation0;

/**
 * An {@code AvailServerChannel} represents an abstract connection between an
 * {@link AvailServer} and a client (represented by a {@link TransportAdapter}).
 * It provides mechanisms for sending and receiving {@linkplain Message
 * messages}.
 *
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 */
public abstract class AvailServerChannel
implements AutoCloseable
{
	/**
	 * Answer the {@linkplain AvailServer server} that created this {@linkplain
	 * AbstractTransportChannel channel}.
	 *
	 * @return The {@code AvailServer} that created this channel.
	 */
	public abstract AvailServer server ();

	/**
	 * Is the {@linkplain AvailServerChannel channel} open?
	 *
	 * @return {@code true} if the channel is open, {@code false} otherwise.
	 */
	public abstract boolean isOpen ();

	/**
	 * Enqueue the given {@linkplain Message message}. When the message has
	 * been enqueued, then execute the {@linkplain Continuation0 continuation}.
	 *
	 * @param message
	 *        A message.
	 * @param enqueueSucceeded
	 *        What to do when the message has been successfully enqueued.
	 */
	public abstract void enqueueMessageThen (
		final Message message,
		final Continuation0 enqueueSucceeded);

	/**
	 * Receive an incoming {@linkplain Message message}.
	 *
	 * @param message
	 *        A message.
	 */
	public abstract void receiveMessage (final Message message);

	/**
	 * {@code ProtocolState} represents the communication state of a {@linkplain
	 * AvailServerChannel server channel}.
	 */
	public static enum ProtocolState
	{
		/** Protocol version must be negotiated. */
		VERSION_NEGOTIATION
		{
			@Override
			Set<ProtocolState> allowedSuccessorStates ()
			{
				return Collections.singleton(ELIGIBLE_FOR_UPGRADE);
			}

			@Override
			public boolean versionNegotiated ()
			{
				return false;
			}
		},

		/**
		 * The {@linkplain AvailServerChannel channel} is eligible for upgrade.
		 */
		ELIGIBLE_FOR_UPGRADE
		{
			@Override
			Set<ProtocolState> allowedSuccessorStates ()
			{
				return new HashSet<>(Arrays.asList(COMMAND, IO));
			}

			@Override
			public boolean eligibleForUpgrade ()
			{
				return true;
			}
		},

		/**
		 * The {@linkplain AvailServerChannel channel} should henceforth be
		 * used to issue {@linkplain CommandMessage commands}.
		 */
		COMMAND
		{
			@Override
			Set<ProtocolState> allowedSuccessorStates ()
			{
				return Collections.emptySet();
			}
		},

		/**
		 * The {@linkplain AvailServerChannel channel} should henceforth be
		 * used for general text I/O.
		 */
		IO
		{
			@Override
			Set<ProtocolState> allowedSuccessorStates ()
			{
				return Collections.emptySet();
			}

			@Override
			public boolean generalTextIO ()
			{
				return true;
			}
		};

		/**
		 * Answer the allowed successor {@linkplain ProtocolState states} of
		 * the receiver.
		 *
		 * @return The allowed successor states.
		 */
		abstract Set<ProtocolState> allowedSuccessorStates ();

		/**
		 * Does this {@linkplain ProtocolState state} indicate that the version
		 * has already been negotiated?
		 *
		 * @return {@code true} if the version has already been negotiated,
		 *         {@code false} otherwise.
		 */
		public boolean versionNegotiated ()
		{
			return true;
		}

		/**
		 * Does this {@linkplain ProtocolState state} indicate eligibility for
		 * upgrade?
		 *
		 * @return {@code true} if the state indicates eligibility for upgrade,
		 *         {@code false} otherwise.
		 */
		public boolean eligibleForUpgrade ()
		{
			return false;
		}

		/**
		 * Does this {@linkplain ProtocolState state} indicate a capability to
		 * do general text I/O?
		 *
		 * @return {@code true} if the state indicates the capability, {@code
		 *         false} otherwise.
		 */
		public boolean generalTextIO ()
		{
			return false;
		}
	}

	/** The current {@linkplain ProtocolState protocol state}. */
	private ProtocolState state = ProtocolState.VERSION_NEGOTIATION;

	/**
	 * Answer the current {@linkplain ProtocolState protocol state}.
	 *
	 * @return The current protocol state.
	 */
	public ProtocolState state ()
	{
		return state;
	}

	/**
	 * Set the {@linkplain ProtocolState protocol state}.
	 *
	 * @param state
	 *        The new protocol state. This must be an {@linkplain
	 *        ProtocolState#allowedSuccessorStates() allowed successor} of the
	 *        {@linkplain #state() current protocol state}.
	 */
	public void setState (final ProtocolState state)
	{
		assert this.state.allowedSuccessorStates().contains(state);
		this.state = state;
	}

	/**
	 * The {@linkplain TextInterface text interface}, or {@code null} if the
	 * {@linkplain AvailServerChannel receiver} is not an upgraded I/O channel.
	 */
	private @Nullable TextInterface textInterface;

	/**
	 * Answer the {@linkplain TextInterface text interface}. This is applicable
	 * for {@linkplain ProtocolState#generalTextIO() general text I/O}
	 * {@linkplain AvailServerChannel channels} only.
	 *
	 * @return The text interface.
	 */
	public TextInterface textInterface ()
	{
		final TextInterface io = textInterface;
		assert io != null;
		return io;
	}

	/**
	 * Upgrade the {@linkplain AvailServerChannel channel} for general I/O.
	 */
	public void upgradeToIOChannel ()
	{
		setState(ProtocolState.IO);
		textInterface = new TextInterface(
			new ServerInputChannel(this),
			new ServerOutputChannel(this),
			new ServerErrorChannel(this));
	}

	/**
	 * The {@link UUID}s of any upgrade requests issued by this {@linkplain
	 * AvailServerChannel channel}.
	 */
	private final Set<UUID> requestedUpgrades = new HashSet<>();

	/**
	 * Record an upgrade request instigated by this {@linkplain
	 * AvailServerChannel channel}.
	 *
	 * @param uuid
	 *        The {@link UUID} that identifies the upgrade request.
	 */
	public void recordUpgradeRequest (final UUID uuid)
	{
		synchronized (requestedUpgrades)
		{
			requestedUpgrades.add(uuid);
		}
	}

	@Override
	protected void finalize ()
	{
		try
		{
			server().discontinueUpgradeRequests(requestedUpgrades);
		}
		catch (final Throwable e)
		{
			// Do not prevent destruction of this channel.
		}
	}

	/**
	 * The next {@linkplain CommandMessage command} {@linkplain AtomicLong
	 * identifier} to issue.
	 */
	private final AtomicLong commandId = new AtomicLong(1);

	/**
	 * Answer the next {@linkplain CommandMessage command identifier} from the
	 * {@linkplain AvailServerChannel channel}'s internal sequence.
	 *
	 * @return The next command identifier.
	 */
	public long nextCommandId ()
	{
		return commandId.getAndIncrement();
	}
}

File: avail-server/src/main/kotlin/com/avail/server/io/AvailServerChannel.kt
/*
 * AvailServerChannel.kt
 * Copyright © 1993-2021, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.server.io

import com.avail.io.TextInterface
import com.avail.server.AvailServer
import com.avail.server.messages.TextCommand
import com.avail.server.messages.CommandMessage
import com.avail.server.messages.Message
import com.avail.server.messages.UpgradeCommandMessage
import com.avail.server.session.Session
import java.util.EnumSet
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong

/**
 * An `AvailServerChannel` represents an abstract connection between an
 * [AvailServer] and a client (represented by a [TransportAdapter]). It provides
 * mechanisms for sending and receiving [messages][Message].
 *
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 * @author Richard Arriaga &lt;rich@availlang.org&gt;
 *
 * @constructor
 * Construct an [AvailServerChannel].
 *
 * @param closeAction
 *   The custom action that is to be called when the this channel is closed in
 *   order to support implementation-specific requirements after the closing of
 *   this channel.
 */
abstract class AvailServerChannel constructor(
	closeAction: (DisconnectReason, AvailServerChannel) -> Unit = {_,_->})
{
	/** `true` if the channel is open, `false` otherwise. */
	abstract val isOpen: Boolean

	/**
	 * The [ChannelCloseHandler] that contains the application-specific close
	 * logic to be run after this [AvailServerChannel] is closed.
	 */
	val channelCloseHandler: ChannelCloseHandler by lazy {
		ChannelCloseHandler(this, closeAction)
	}

	/**
	 * The unique channel [identifier][UUID] used to distinguish this
	 * [AvailServerChannel] from other `AvailServerChannel`s.
	 *
	 * Only command channels will keep their UUID assigned upon construction.
	 * All subordinate IO-channels will have their identifiers set to the
	 * [TextCommand.UPGRADE] token received from a client-sent
	 * [UpgradeCommandMessage].
	 */
	var id: UUID = UUID.randomUUID()

	/**
	 * The [id] of the command [channel][AvailServerChannel] that is responsible
	 * for spawning this channel if this channel was created through the
	 * [upgrade][TextCommand.UPGRADE] process.
	 *
	 * If the value is `null` that means that either one of two things has
	 * happened:
	 *
	 *  1. This is a new channel that has not been upgraded to a specific type
	 *     (e.g IO, Binary, or COMMAND).
	 *  2. This is a COMMAND channel (ProtocolState = COMMAND)
	 *
	 * If this parent id is not null, then the parent id is the same as the
	 * [Session.id] to which this channel belongs to.
	 */
	var parentId: UUID? = null

	/**
	 * The [Session.id] of the [Session] this [AvailServerChannel] belongs to;
	 * `null` if this `AvailServerChannel` has not yet been assigned to a
	 * session.
	 */
	val sessionId: UUID? get() =
		when
		{
			state == ProtocolState.COMMAND -> id
			parentId != null -> parentId
			else -> null
		}

	/**
	 * The [Session] this [AvailServerChannel] belongs to or `null` if not yet
	 * assigned to a `Session`.
 	 */
	val session: Session? get() = sessionId?.let {
		server.sessions[it]
	}

	/**
	 * The time in milliseconds since the Unix Epoch when this
	 * [AvailServerChannel] was first created.
	 */
	val created = System.currentTimeMillis()

	/**
	 * `true` if this [AvailServerChannel] is an IO channel for a
	 * [parent][parentId] command channel.
	 */
	private val isIOChannel get() = state.generalTextIO

	/** The current [protocol state][ProtocolState].  */
	var state = ProtocolState.VERSION_NEGOTIATION
		set (newState)
		{
			assert(this.state.allowedSuccessorStates.contains(newState))
			field = newState
		}

	/**
	 * Schedule this [AvailServerChannel] to be closed cleanly if possible. This
	 * does not guarantee that the channel will close right away nor does it
	 * guarantee that it will be closed cleanly.
	 *
	 * @param reason
	 *   The [DisconnectReason] for closing the channel.
	 */
	abstract fun scheduleClose (reason: DisconnectReason)

	/**
	 * Close this [AvailServerChannel] immediately. The provided reason will
	 * be provided to the [channelCloseHandler] as the closing reason,
	 * overriding any previously stated close reason if any had been provided
	 * during a [scheduleClose] request.
	 *
	 * @param reason
	 *   The [DisconnectReason] for closing the channel.
	 */
	abstract fun closeImmediately (reason: DisconnectReason)

	/**
	 * The [text&#32;interface][TextInterface], or `null` if the
	 * [receiver][AvailServerChannel] is not an upgraded I/O channel.
	 */
	var textInterface: TextInterface? = null

	/**
	 * A `ServerInputNotificationChannel` adapts a [channel][AvailServerChannel]
	 * for use as a notifier of a pending interest in receiving standard input.
	 */
	inner class ServerInputNotificationChannel
		: AbstractServerOutputChannel(this)
	{
		override val channelTag = "in"
	}

	/**
	 * The [ServerInputNotificationChannel], if any. Present only for I/O
	 * channels.
	 */
	var inputNotificationChannel: ServerInputNotificationChannel? = null

	/**
	 * The [UUID]s of any upgrade requests issued by this
	 * [channel][AvailServerChannel].
	 */
	private val requestedUpgrades = mutableSetOf<UUID>()

	/** The next [command][CommandMessage] [identifier][AtomicLong] to issue. */
	private val commandId = AtomicLong(1)

	/**
	 * The [server][AvailServer] that created this
	 * [channel][AbstractTransportChannel].
	 */
	abstract val server: AvailServer

	/**
	 * Enqueue the given [message][Message]. When the message has been enqueued,
	 * then execute the action.
	 *
	 * @param message
	 *   A message.
	 * @param enqueueSucceeded
	 *   What to do when the message has been successfully enqueued.
	 */
	abstract fun enqueueMessageThen(
		message: Message,
		enqueueSucceeded: ()->Unit)

	/**
	 * Receive an incoming [message][Message].
	 *
	 * @param message
	 *   A message.
	 */
	abstract fun receiveMessage(message: Message)

	/**
	 * `ProtocolState` represents the communication state of a
	 * [server&#32;channel][AvailServerChannel].
	 *
	 * @author Todd L Smith &lt;todd@availlang.org&gt;
	 */
	enum class ProtocolState
	{
		/** Protocol version must be negotiated.  */
		VERSION_NEGOTIATION
		{
			override val allowedSuccessorStates
				get() = setOf(ELIGIBLE_FOR_UPGRADE)
			override val versionNegotiated get() = false
		},

		/**
		 * The [channel][AvailServerChannel] is eligible for upgrade.
		 */
		ELIGIBLE_FOR_UPGRADE
		{
			override val allowedSuccessorStates: Set<ProtocolState>
				get() = EnumSet.of(COMMAND, IO, BINARY)
			override val eligibleForUpgrade get() = true
		},

		/**
		 * The [channel][AvailServerChannel] should henceforth be used to issue
		 * [commands][CommandMessage].
		 */
		COMMAND
		{
			override val allowedSuccessorStates: Set<ProtocolState>
				get() = emptySet()
		},

		/**
		 * The [channel][AvailServerChannel] should henceforth be used for
		 * general text I/O.
		 */
		IO
		{
			override val allowedSuccessorStates: Set<ProtocolState>
				get() = emptySet()
			override val generalTextIO get() = true
		},

		/**
		 * The [channel][AvailServerChannel] should henceforth be used for
		 * general binary data transmission.
		 */
		BINARY
		{
			override val allowedSuccessorStates: Set<ProtocolState>
				get() = emptySet()
			override val generalBinary get() = true
		};

		/** The allowed successor [states][ProtocolState] of the receiver. */
		internal abstract val allowedSuccessorStates: Set<ProtocolState>

		/**
		 * Does this [state][ProtocolState] indicate that the version has
		 * already been negotiated? `true` if the version has already been
		 * negotiated, `false` otherwise.
		 */
		open val versionNegotiated get() = true

		/**
		 * Does this [state][ProtocolState] indicate eligibility for upgrade?
		 * `true` if the state indicates eligibility for upgrade, `false`
		 * otherwise.
		 */
		open val eligibleForUpgrade get() = false

		/**
		 * Does this [state][ProtocolState] indicate a capability to do general
		 * text I/O? `true` if the state indicates the capability, `false`
		 * otherwise.
		 */
		open val generalTextIO get() = false

		/**
		 * Does this [state][ProtocolState] indicate a capability to transmit
		 * binary? `true` if the state indicates the capability, `false`
		 * otherwise.
		 */
		open val generalBinary get() = false
	}

	/**
	 * Upgrade the [channel][AvailServerChannel] for general text I/O.
	 */
	fun upgradeToIOChannel()
	{
		state = ProtocolState.IO
		textInterface = TextInterface(
			ServerInputChannel(this),
			ServerOutputChannel(this),
			ServerErrorChannel(this))
		inputNotificationChannel = ServerInputNotificationChannel()
	}

	/**
	 * Record an upgrade request instigated by this
	 * [channel][AvailServerChannel].
	 *
	 * @param uuid
	 *   The [UUID] that identifies the upgrade request.
	 */
	fun recordUpgradeRequest(uuid: UUID)
	{
		synchronized(requestedUpgrades) {
			requestedUpgrades.add(uuid)
		}
	}

	protected fun finalize()
	{
		try
		{
			server.discontinueUpgradeRequests(requestedUpgrades)
		}
		catch (e: Throwable)
		{
			// Do not prevent destruction of this channel.
		}

	}

	/**
	 * The next [command&#32;identifier][CommandMessage] from the
	 * [channel][AvailServerChannel]'s internal sequence.
	 */
	val nextCommandId get() = commandId.getAndIncrement()

	override fun toString(): String =
		if (isIOChannel)
		{
			"($id) $state: [$parentId] "
		}
		else
		{
			"($id) $state"
		}
}
